Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +26,63 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+       return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int ms_remaining;
+      if (isCharging() == true)
+      {
+        ms_remaining = totalAmpHoursUsed() / averageCurrent_ * MS_CONVERSION;
+      }   else
+      {
+        ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * MS_CONVERSION;
+      }
+
+    QTime timeWhenChargedOrDepleted = time.addMSecs(ms_remaining);
+
+    return timeWhenChargedOrDepleted;
+
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    int deltaMseconds;
+    double chargeAh;
+    QTime msTime(0,0,0,0);
+
+    if (previousMilliseconds_.isNull())
+    {
+        previousMilliseconds_ = initialTime_.addMSecs(batteryData.time.msec());
+    }
+
+    msTime = msTime.addMSecs(batteryData.time.msec());
+
+    deltaMseconds = abs(msTime.msecsTo(previousMilliseconds_));
+
+    double deltaHours = deltaMseconds * HOUR_CONVERSION;
+
+    previousMilliseconds_ = initialTime_.addMSecs(batteryData.time.msec());
+
+    current_ = batteryData.current;
+
+    totalCurrent_ = batteryData.current + totalCurrent_;
+
+    averageCurrent_ = totalCurrent_ / counter_;
+
+    counter_++;
```

Comment:
Put this increment at the beginning, and set the initial value to zero. While functionally the same, it makes more sense to increment the counter of data only when new data is coming in, not before.